File: neetcode/blind75/backtracking/combination_target_sum.py
# https://neetcode.io/problems/combination-target-sum

class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []

        if (not nums):
            return res

        # idx: always start combo from current idx
        # curr_sum: current value needed to equate to target
        # curr_combo: current combination of num attempting to equate to target.
        def backtracking(idx, curr_sum, curr_combo):
            if (curr_sum < 0):
                return
            elif (curr_sum == 0):
                res.append(curr_combo.copy())
            
            for i in range(idx, len(nums)):
                # Add to potential solution combo
                curr_combo.append(nums[i])
                # eval the combo
                backtracking(i, curr_sum - nums[i], curr_combo);
                # remove for next number
                curr_combo.pop()

            return

        backtracking(0, target, [])
        print(res)
        return res

File: neetcode/neet150_JS/Backtracking/combination-target-sum.js
// https://neetcode.io/problems/combination-target-sum

class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {
        if (nums.length === 0) {
            return []
        }

        const res = []
        const subset = []

        this.DFS(nums, target, 0, subset, res)
        return res
    }

    DFS(nums, target, i, subset, res) {
        if (target === 0) {
            res.push(Array.from(subset))
            return
        }
        if (i >= nums.length || target < 0) {
            return
        }

        subset.push(nums[i])
        this.DFS(nums, target - nums[i], i, subset, res)

        subset.pop()
        this.DFS(nums, target, i + 1, subset, res)

        return
    }
}
